src/components/timeline/TrackScreenshots.js

```diff
@@ -0,0 +1,250 @@
+    const { screenshots, thread, isMakingPreviewSelection, width } = this.props;
+    if (isMakingPreviewSelection || offsetX === null || pageX === null) {
+      return null;
+    }
+    const screenshotIndex = this.findScreenshotAtMouse(offsetX);
+    if (screenshotIndex === null) {
+      return null;
+    }
+    // Coerce the payload into a screenshot one.
+    const payload: ScreenshotPayload = (screenshots[screenshotIndex].data: any);
+    const { url, windowWidth, windowHeight } = payload;
+
+    // Compute the hover image's thumbnail size.
+    let hoverHeight = HOVER_HEIGHT;
+    let hoverWidth = HOVER_HEIGHT / windowHeight * windowWidth;
+
+    if (hoverWidth > HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO) {
+      // This is a really wide image, limit the height so it lays out reasonably.
+      hoverWidth = HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO;
+      hoverHeight = hoverWidth / windowWidth * windowHeight;
+    }
+
+    // Set the top so it centers around the track.
+    const top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
+    const left =
+      offsetX + hoverWidth * 0.5 > width
+        ? // Stick the hover image on to the right side of the container.
+          pageX - offsetX + width - hoverWidth * 0.5
+        : // Center the hover image around the mouse.
+          pageX;
+
+    return createPortal(
+      <div className="timelineTrackScreenshotHover" style={{ left, top }}>
+        <img
+          className="timelineTrackScreenshotHoverImg"
+          src={thread.stringTable.getString(url)}
+          style={{
+            height: hoverHeight,
+            width: hoverWidth,
+          }}
+        />
+      </div>,
+      this._overlayElement
+    );
+  }
+
+  _handleMouseLeave = () => {
+    this.setState({
+      offsetX: null,
+      pageX: null,
+      containerTop: null,
+    });
+  };
+
+  _handleMouseMove = (event: SyntheticMouseEvent<HTMLDivElement>) => {
+    const { top, left } = event.currentTarget.getBoundingClientRect();
+    this.setState({
+      pageX: event.pageX,
+      offsetX: event.pageX - left,
+      containerTop: top,
+    });
+  };
+
+  render() {
+    return (
+      <div
+        className="timelineTrackScreenshot"
+        style={{ height: TRACK_HEIGHT }}
+        onMouseLeave={this._handleMouseLeave}
+        onMouseMove={this._handleMouseMove}
+      >
+        {this.renderScreenshotStrip()}
+        {this.renderHoverPreview()}
+      </div>
+    );
+  }
+}
+
+const options: ExplicitConnectOptions<OwnProps, StateProps, DispatchProps> = {
+  mapStateToProps: (state, ownProps) => {
+    const { threadIndex, screenshotId } = ownProps;
+    const selectors = selectorsForThread(threadIndex);
+    const { start, end } = getCommittedRange(state);
+    const previewSelection = getPreviewSelection(state);
+    return {
+      thread: selectors.getRangeFilteredThread(state),
+      screenshots: ensureExists(
+        selectors.getRangeFilteredScreenshotsById(state).get(screenshotId),
```

Comment:
I think that the fact that it won't be null is now OK. But this relies on a lot of somewhat implicit assumptions in 1. how we compute the global tracks (the rule that there won't be a screenshot track if there are no screenshot markers), and 2. how we compute this selector (the rule that if there's at least a screenshot marker for this id in the thread, there will always be a non-null result).

Therefore, if you want to keep this `ensureExists` here, I'd like some solid tests to make sure we don't regress in the future.
For 2. you wrote a test already so it's good (the test where we reduce the range).
For 1. I'd like that you check if such a test exists: merely a full snapshot test for a profile that doesn't have any screenshot markers. Maybe we have one already. Maybe it makes sense to explicitely add one with the proper title even if we implicitely test it already.

Alternatively, you can make sure to pass an empty array as `screenshots` if the map doesn't have this ID.